dot/state/grandpa.go

```diff
@@ -24,44 +26,570 @@ var (
+		return nil, fmt.Errorf("cannot set change set id at block 0: %w", err)
 	}
 
 	return s, nil
 }
 
 // NewGrandpaState returns a new GrandpaState
-func NewGrandpaState(db chaindb.Database) (*GrandpaState, error) {
+func NewGrandpaState(db chaindb.Database, bs *BlockState) (*GrandpaState, error) {
 	return &GrandpaState{
-		db: chaindb.NewTable(db, grandpaPrefix),
+		db:         chaindb.NewTable(db, grandpaPrefix),
+		blockState: bs,
 	}, nil
 }
 
+// HandleGRANDPADigest receives a decoded GRANDPA digest and calls the right function to handles the digest
+func (s *GrandpaState) HandleGRANDPADigest(header *types.Header, digest scale.VaryingDataType) error {
+	switch val := digest.Value().(type) {
+	case types.GrandpaScheduledChange:
+		return s.addScheduledChange(header, val)
+	case types.GrandpaForcedChange:
+		return s.addForcedChange(header, val)
+	case types.GrandpaOnDisabled:
+		return nil
+	case types.GrandpaPause:
+		return nil
+	case types.GrandpaResume:
+		return nil
+	default:
+		return fmt.Errorf("not supported digest")
+	}
+}
+
+func (s *GrandpaState) addForcedChange(header *types.Header, fc types.GrandpaForcedChange) error {
+	headerHash := header.Hash()
+
+	for _, change := range s.forcedChanges {
+		changeBlockHash := change.announcingHeader.Hash()
+
+		if changeBlockHash.Equal(headerHash) {
+			return errDuplicateHashes
+		}
+
+		isDescendant, err := s.blockState.IsDescendantOf(changeBlockHash, headerHash)
+		if err != nil {
+			return fmt.Errorf("cannot verify ancestry: %w", err)
+		}
+
+		if isDescendant {
+			return errAlreadyHasForcedChanges
+		}
+	}
+
+	auths, err := types.GrandpaAuthoritiesRawToAuthorities(fc.Auths)
+	if err != nil {
+		return fmt.Errorf("cannot parser GRANDPA authorities to raw authorities: %w", err)
+	}
+
+	pendingChange := &pendingChange{
+		bestFinalizedNumber: fc.BestFinalizedBlock,
+		nextAuthorities:     auths,
+		announcingHeader:    header,
+		delay:               fc.Delay,
+	}
+
+	s.forcedChanges = append(s.forcedChanges, pendingChange)
+	sort.Sort(s.forcedChanges)
+
+	logger.Debugf("there are now %d possible forced changes", len(s.forcedChanges))
+	return nil
+}
+
+func (s *GrandpaState) addScheduledChange(header *types.Header, sc types.GrandpaScheduledChange) error {
+	auths, err := types.GrandpaAuthoritiesRawToAuthorities(sc.Auths)
+	if err != nil {
+		return fmt.Errorf("cannot parse GRANPDA authorities to raw authorities: %w", err)
+	}
+
+	pendingChange := &pendingChange{
+		nextAuthorities:  auths,
+		announcingHeader: header,
+		delay:            sc.Delay,
+	}
+
+	return s.importScheduledChange(pendingChange)
+}
+
+func (s *GrandpaState) importScheduledChange(pendingChange *pendingChange) error {
+	highestFinalizedHeader, err := s.blockState.GetHighestFinalisedHeader()
+	if err != nil {
+		return fmt.Errorf("cannot get highest finalized header: %w", err)
+	}
+
+	if pendingChange.announcingHeader.Number <= highestFinalizedHeader.Number {
+		return errLowerThanBestFinalized
+	}
```

Comment:
won't this reject any changes that are in unfinalized blocks? but I think this function is called for imported (unfinalized) blocks, right?

Reply:
Yes! Every time a block is imported we check the digest for scheduled changes to import and If the node has a finalized block n so we should not import a scheduled change from blocks before n but we should import changes from blocks ahead of n, right?

Reply:
oh hmm, I think this check can probably be removed, because we might be syncing somewhat slowly so by the time we sync a block with a grandpa change digest, it could be finalized already, but we should still take it into account